Fix some issues around merge key/anchor handling

stevenwdv opened the pull request:
This PR fixes multiple issues related to merge keys (`!!merge <<`):

- Allow inline maps instead of just aliases (e.g. `{<<: {a: 42}}`, `{<<: [{a: 42}]}`), both for traversing and exploding (closes #2386)
- Allow aliases to sequences to be used (e.g. `{s: &s [{a: 42}], m: {<<: *s}}`), both for traversing and exploding (closes #2178)
- Disallow nested sequences (e.g. ~~`{<<: [[{a: 42}]]}`~~)
- Disallow other (inline) types (e.g. ~~`{<<: 42}`~~)
- However: Ignore invalid merge keys when traversing maps. Previously cases like `{a: &a 42, m: {<<: *a}}` were reported but cases like `{m: {<<: 42}}` were not. I chose to ignore all errors because otherwise traversing such a map (e.g. to correct types) becomes impossible.
- Errors are *not* ignored when using `explode`. Previously explode would just leave out invalid merge keys. Another potential solution would be to leave them as-is.
- Fix precedence of merge anchor sequence for traverse (explode was already correct), now it correctly lets earlier keys overwrite later keys according to the spec.
- Fix excessive exploding with merge keys, e.g. previously `echo '{b: &b {a: &a 42}, r: *a, c: {<<: *b}}' | yq 'explode(.c)'` would yield `{b: &b {a: 42}, r: *a, c: {a: 42}}`, making `r: *a` a dangling reference. Now it yields `{b: &b {a: &a 42}, r: *a, c: {a: &a 42}}`.
- For exploding, detect merge keys using tag (`!!merge`) rather than value (`<<`).

I do still have a question: I saw in `operator_traverse_path_test.go` that goccy would reject at least some invalid merge types at parse time, which might make using custom tags like `<<: !include file.yaml` (see #2386) impossible. Am I right about that? And does that mean that this usage will not be supported in the future? If so, that would be a bit sad. See also the note above about ignoring invalid merge keys.

stevenwdv commented on pkg/yqlib/operator_traverse_path_test.go:
```diff
@@ -553,9 +569,16 @@ var traversePathOperatorScenarios = []expressionScenario{
+		document:      `{<<: 42}`,
+		expression:    ".[]",
+		expectedError: "can only use merge anchors with maps (!!map) or sequences (!!seq) of maps, but got !!int",
+		skipForGoccy:  true,
```
I didn't try to test goccy but I assume it will reject this if it rejects the case above

stevenwdv commented:
Wait, this actually doesn't fix `explode(.)` yet. Converting to draft for now, I'll do that later.

mikefarah commented on pkg/yqlib/operator_anchors_aliases_test.go:
```diff
@@ -283,6 +283,66 @@ var anchorOperatorScenarios = []expressionScenario{
 		expression:     `.thingOne |= explode(.) * {"value": false}`,
 		expected:       []string{expectedUpdatedArrayRef},
 	},
+	{ // Merge anchor with inline map
```
put comments as the description instead, they will then be printed on test failure for traceability

mikefarah commented:
Minor comment, need more time to review this and consider if there are any side effects - thanks for the contribution though!

mikefarah commented on pkg/yqlib/operator_traverse_path_test.go:
```diff
@@ -425,11 +449,11 @@ var traversePathOperatorScenarios = []expressionScenario{
 		expression:     `.foobarList.thing`,
 		expected: []string{
-			"D0, P[bar thing], (!!str)::bar_thing\n",
+			"D0, P[foo thing], (!!str)::foo_thing\n",
```
Oooh nice find - this isn't to the YAML spec and it's works opposite to explode 🤦🏼  _but_ this is a breaking change... and may cause all sorts of issues for people :/ Hmmmm

mikefarah replied:
I'm happy to go through with this PR if either; this fix is not included, or it's controlled via a new flag, maybe `--yaml-correct-merge-order` or something, defaulting to false. Should have a warning in there if the flag is not set and the logic is executed.

Then later with enough notice given, I can update the default flag value to true.

mikefarah replied:
Yeah good idea 👍🏼

mikefarah replied:
I think it could be done under the same flag, I'm happy to work on 2110 and adding a flag, then perhaps you can add the `if` check into this PR?

stevenwdv replied:
Nice! I put the fixed overriding behind the flag. You missed fixing #2110 for traversing (only exploding), but I could fix it both for local keys (#2110) & list merge keys at once anyway.

I also noticed that after 6e8cc00030b8165f159d72cc39ccfd67a97c483d it'll warn about overriding behavior with list merge anchor keys as well (e.g. `{<<: [{a: 1}, {a: 2}]}`), while this behavior is and was already correct (`explode(.).a` = `1`), see the `TODO`s I added. Maybe you know of a good fix? Or add 'unless...' to the warning..?

mikefarah commented on pkg/yqlib/operator_anchors_aliases_test.go:
```diff
@@ -122,12 +142,22 @@ var anchorOperatorScenarios = []expressionScenario{
 		expression:     ".[4] | explode(.)",
 		expected:       []string{"D0, P[4], (!!map)::r: 10\nx: 1\ny: 2\n"},
 	},
+	//TODO The following 2 tests warn about overwriting [3].r not being to spec while they shouldn't
```
Ah I see...thanks for making this super clear! Will fix...somehow

mikefarah commented:
This is really good - I think you have a clearer picture of how merge works better than I do 😅  - I'm going to go over it again later just to double check it as it's a sensitive part of the code.

stevenwdv commented:
Thanks for having another look! Did you see my comments in the [thread above](https://github.com/mikefarah/yq/pull/2400#discussion_r2210545616) as well? Currently I've actually broken key overriding in regular maps for traversing, and I don't have a good solution yet...

stevenwdv commented:
@mikefarah I notice that `fixedReconstructAliasedMap` in `MakeExplodeGreatAgain` is lacking a lot of logic that is present in `reconstructAliasedMap`. Compare:
```
$ echo '{e: &empty {}, b: &b 42, a: {b2: *b}, <<: *empty}' | ./yq 'explode(.)'
[WARNING] 14:18:06 --yaml-fix-merge-anchor-to-spec is false; causing merge anchors to override the existing values which isn't to the yaml spec. This flag will default to true in late 2025.
{e: {}, b: 42, a: {b2: 42}}
```
vs
```
$ echo '{e: &empty {}, b: &b 42, a: {b2: *b}, <<: *empty}' | ./yq --yaml-fix-merge-anchor-to-spec 'explode(.)'
{e: &empty {}, b: &b 42, a: {b2: *b}}
```
This is because it's not using `overrideEntry`. It's also not using `Context`, but I'm actually not sure what the exact purpose of that is.

stevenwdv commented:
> I'm seeing if I can quickly fix it in the merge.

I thought of a way to quickly solve the issue in a different way, by evaluating merge keys first, I hope you don't mind.

mikefarah commented:
Ah the reason I was processing like that was to preserve key order on merging. Not a big issue though, this is looking good! Thanks for PR!

mikefarah commented:
Hmm - only problem with the way you're processing explode is that they key order is mangled if merge is processed first :/ Not a massive issues I guess...though I might see if I can fix that

stevenwdv commented:
Hey @mikefarah, I'm really confused because it seems like something went really 
wrong merging this; a lot of changes didn't make it and some logic from `MakeExplodeGreatAgain` was merged while I changed it in this MR. E.g.
- `fixedReconstructAliasedMap` re-appeared
- `!!str <<` is broken again
- Inline merge anchors are broken without `--yaml-correct-merge-order` I fixed those for both cases
- Nested merge anchors are broken even with the fix
- Probably more

mikefarah commented:
Yeah, I brought those changes back in to ensure the correct map key order. Afaik I included all the tests you added; so there's shouldn't be any gaps :/ - but maybe something went awry for those cases. 

Inline merging is under the flag, that's intentional - my thinking is that it's best to leave the old way of merging as untouched as possible seeing the new way is a complete re-write.

Not sure what `!!str <<` exactly refers to? Will take a look at nested inline merge anchors when I get a chance - though it seems pretty niche; not sure what the use-case would be?

stevenwdv commented:
@mikefarah
I'm a bit sad to see many things that worked with this MR not work anymore. I put in a lot of work to try and get everything right, only to see most of it break again. Sorry if that sounds to harsh, I'm still really glad you're maintaining this useful tool and actively looking at issues and comments. Anyway, more specifically, it still doesn't work anymore for what I was trying to do originally, also with the flag. (Specifically because of nested inline merge anchors.)

I think that aside from the order, which I understand you might want to keep the same but shouldn't impact the meaning, I made sure to keep everything compatible with previous behavior with and without the flag. I also think that I kept the order the same as before when the flag was not specified (although I'm not 100% sure of the top of my head); maybe you can check this, and in that case, can't the other fixes just be included with and without the flag? I don't think those could break previous code.

Regarding behavior with the flag, `fixedReconstructAliasedMap` bypasses all original fixes from this MR, so it doesn't handle inline maps in merge anchors, and has some other shortcomings.

I think that not all tests are ran for both with & without the flag anymore where they could be. Maybe that's why it seems to be working fine.

`!!str <<` is not a merge key, it's just a regular key that happens to be called `<<`.

I could make a PR to fix the things that are broken now, but the thing is; this was that PR :/
